src/Timer/Timer.c

```diff
@@ -0,0 +1,381 @@
+            break;
+        case Timer_Id_2:
+            result = (Timer_Apbctrl1_Registers) Timer2_Apbctrl1_Address;
+            break;
+        case Timer_Id_3:
+            result = (Timer_Apbctrl1_Registers) Timer3_Apbctrl1_Address;
+            break;
+        case Timer_Id_4:
+            result = (Timer_Apbctrl1_Registers) Timer4_Apbctrl1_Address;
+            break;
+        default:
+            result = (Timer_Apbctrl1_Registers) TimerInvalid_Apbctrl1_Address;
+    }
+
+    return result;
+}
+
+void
+Timer_Apbctrl1_init(Timer_Id id, Timer_Apbctrl1 *const timer, const Timer_InterruptHandler handler)
+{
+#ifdef MOCK_REGISTERS
+    baseApbctrl1Registers = malloc(sizeof(Timer_Apbctrl1_Base_Registers));
+    timer->regs = malloc(sizeof(Timer_Apbctrl1_Registers));
+#else
+    baseApbctrl1Registers = (Timer_Apbctrl1_Base_Registers) GPTIMER_APBCTRL1_ADDRESS_BASE;
+    timer->regs = getApbctrl1TimerAddressById (id);
+#endif
+    Timer_baseInit (&baseApbctrl1Registers->configuration);
+    timer->id = id;
+    timer->irqHandler = handler;
+    timer->rtemsInterruptEntry = (rtems_interrupt_entry){0};
+    timer->regs->control = 0;
+    timer->regs->counter = 0;
+    timer->regs->reload = 0;
+
+    uint8_t irqNumber = Timer_getApbctrl1InterruptNumber(timer->id);
+    irqInit(&timer->rtemsInterruptEntry, &timer->irqHandler, irqNumber);
+}
+
+void
+Timer_Apbctrl1_setBaseScalerReloadValue(uint16_t scalerReloadValue)
+{
+    baseApbctrl1Registers->reload = (uint32_t) scalerReloadValue;
+}
+
+void
+Timer_Apbctrl1_setConfig(Timer_Apbctrl1 *const timer, const Timer_Config *const config)
+{
+    rtems_interrupt_vector_disable(Timer_getApbctrl1InterruptNumber(timer->id));
+    Timer_setConfig (&timer->regs->control, &timer->regs->reload, config);
+    rtems_interrupt_vector_enable(Timer_getApbctrl1InterruptNumber(timer->id));
+}
+
+void
+Timer_Apbctrl1_getConfig(const Timer_Apbctrl1 *const timer, Timer_Config *const config)
+{
+    Timer_getConfig (timer->regs->control, timer->regs->reload, config);
+}
+
+void
+Timer_Apbctrl1_start(Timer_Apbctrl1* const timer)
+{
+    rtems_interrupt_vector_disable(Timer_getApbctrl1InterruptNumber(timer->id));
+    Timer_start (&timer->regs->control);
+    rtems_interrupt_vector_enable(Timer_getApbctrl1InterruptNumber(timer->id));
+}
+
+void
+Timer_Apbctrl1_stop(Timer_Apbctrl1 *const timer)
+{
+    rtems_interrupt_vector_disable(Timer_getApbctrl1InterruptNumber(timer->id));
+    Timer_stop (&timer->regs->control);
+}
+
+void
+Timer_Apbctrl1_shutdown(Timer_Apbctrl1 *const timer)
+{
+    rtems_interrupt_vector_disable(Timer_getApbctrl1InterruptNumber(timer->id));
+    Timer_stop (&timer->regs->control);
+    irqDeinit(&timer->rtemsInterruptEntry, Timer_getApbctrl1InterruptNumber(timer->id));
+}
+
+uint32_t
+Timer_Apbctrl1_getCounterValue(const Timer_Apbctrl1 *const timer)
+{
+    uint32_t result = 0;
+
+    rtems_interrupt_vector_disable(Timer_getApbctrl1InterruptNumber(timer->id));
```

Comment:
Are we positive that interrupt needs to be disabled to get the counter value?

Reply:
ack